### PR TITLE
Add payment check methods to registration

### PR DESCRIPTION
**Pull request by irisfaraway**
https://eaflood.atlassian.net/browse/RUBY-704

This will let us check for unpaid balances on registrations as well as renewals.

### CODE REVIEW

File: app/models/waste_carriers_engine/registration.rb [hunk 1]
```diff
@@ -30,6 +30,10 @@ def can_start_renewal?
       renewable_tier? && renewable_status? && renewable_date?
     end
 
+    def pending_payment?
+      unpaid_balance?
```
Comment (cintamani) [inline app/models/waste_carriers_engine/registration.rb]: can we use `alias` here?

File: app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb [hunk 2]
```diff
@@ -157,6 +157,10 @@ def update_last_modified
       def declaration_confirmed?
         declaration == 1
       end
+
+      def unpaid_balance?
+        finance_details.present? && finance_details.balance.positive?
```
Comment (cintamani) [inline app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb]: [NIT]
```suggestion
        finance_details.presence&.balance.positive?
```

File: spec/models/waste_carriers_engine/registration_spec.rb [hunk 3]
```diff
@@ -566,6 +566,40 @@ module WasteCarriersEngine
       end
     end
 
+    describe "#pending_payment?" do
```
Comment (cintamani) [inline spec/models/waste_carriers_engine/registration_spec.rb]: if we `alias`, we won't need to test this, instead we should test `unpaid_balance?`. Also, this test is not a pure unit, shall we still go for it?

**Comment by irisfaraway** [comment]
Updated after review comments.

I realise that the "Can have registration attributes" shared examples are currently only testing transient_registrations 🤦‍♀ I'm fixing this in another branch at the moment though so hopefully it won't be a blocker to merging this.

File: app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb [hunk 4]
```diff
@@ -157,6 +157,12 @@ def update_last_modified
       def declaration_confirmed?
         declaration == 1
       end
+
+      def unpaid_balance?
+        return true if finance_details.presence&.balance&.positive?
```
Comment (cintamani) [inline app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb]: Just a question: the previous version of the code always assumed a finance_details, if present, should always have a `balance` present that has a numeric value/object. Is this not the case any more?

File: app/models/waste_carriers_engine/transient_registration.rb [hunk 5]
```diff
@@ -102,7 +102,7 @@ def renewal_application_submitted?
     end
 
     def pending_payment?
```
Comment (cintamani) [inline app/models/waste_carriers_engine/transient_registration.rb]: I was expecting to see the spec for this method to change as well with an `expect(subject).to receive(:unpaid_balance?) ....`, shall we update it?

File: app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb (hunk 4 above)
Comment (irisfaraway) [reply app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb]: It should have a `balance` (although MongoDB is less strict about this sort of thing). But Rubocop kicked off when I had `finance_details.presence&.balance.positive?` – I think it doesn't like to have `.` calls chained after `&.` calls.

Comment (cintamani) [reply app/models/concerns/waste_carriers_engine/can_have_registration_attributes.rb]: Thanks for the explanation! :)

File: app/models/waste_carriers_engine/transient_registration.rb (hunk 5 above)
Comment (irisfaraway) [reply app/models/waste_carriers_engine/transient_registration.rb]: Fixed: https://github.com/DEFRA/waste-carriers-engine/pull/534/commits/b4d267b38f72e90177cfd7b70fb39df567b932d7

**Pull request merged**